nxcomp/Log.cpp

```diff
@@ -0,0 +1,126 @@
+/*                                                                        */
+/* Redistribution and use of the present software is allowed according    */
+/* to terms specified in the file LICENSE.nxcomp which comes in the       */
+/* source distribution.                                                   */
+/*                                                                        */
+/* All rights reserved.                                                   */
+/*                                                                        */
+/* NOTE: This software has received contributions from various other      */
+/* contributors, only the core maintainers and supporters are listed as   */
+/* copyright holders. Please contact us, if you feel you should be listed */
+/* as copyright holder, as well.                                          */
+/*                                                                        */
+/**************************************************************************/
+
+
+#include <stdio.h>
+#include <iostream>
+#include <fstream>
+#include <sstream>
+#include <iomanip>
+
+#include "Log.h"
+
+NXLog nx_log;
+
+
+bool NXLog::will_log() const
+{
+	std::map<std::string, NXLogLevel>::const_iterator item = per_file_levels_.find(current_file());
+
+	if ( item != per_file_levels_.end() )
+	{
+		return current_level() >= item->second;
+	}
+	else
+	{
+		return current_level() >= level();
+	}
+}
+
+
+std::string NXLog::stamp_to_string(const NXLogStamp& stamp) const
+{
+	std::ostringstream oss;
+
+	static const char* level_names[] = {
+		"DEBUG",
+		"INFO ",
+		"WARN ",
+		"ERROR",
+		"FATAL"
+	};
+
+	if ( log_level() )
+		oss << ((stamp.level() >=0 && stamp.level() < NXLOG_LEVEL_COUNT ) ? level_names[stamp.level()] : "???") << " ";
+
+	if ( log_time() )
+	{
+		struct timeval timestamp = stamp.timestamp();
+		struct tm timeinfo;
+
+		localtime_r(&timestamp.tv_sec, &timeinfo);
+
+		if ( log_unix_time() )
+		{
+			oss << timestamp.tv_sec;
+		}
+		else
+		{
+			#if __cplusplus >= 201103L 
+			oss << " " << std::put_time(&timeinfo, "%Y/%m/%d %H:%M:%S");
+			#else
+			oss << timestamp.tv_sec;
+			#endif
+		}
+
+		oss << "." << std::setw(3) << std::setfill('0') << (int)(timestamp.tv_usec / 1000) << " ";
+	}
+
+	if ( log_location() )
+		oss << stamp.file() << "/" << stamp.function() << ":" << stamp.line()  << " ";
+
+	if ( log_thread_id() )
+	{
+		if ( thread_name().empty() )
+			oss << pthread_self() << " ";
+		else
+			oss << "[" << thread_name() << "] ";
+	}
+
+	return oss.str();
+}
+
+NXLog& operator<< (NXLog& out, const NXLogStamp& value)
+{
+	out.current_level( value.level() );
+	out.current_file( value.file() );
+
+	if (  out.will_log() )
+	{
+		if ( out.synchronized() )
+		{
+			out.flush(); // Flush the previous line
+			NXLog::per_thread_data *pdt = out.get_data();
+			(*pdt->buffer) << out.stamp_to_string(value);
+		}
+		else
+		{
+			*(out.stream()) << out.stamp_to_string(value) << " ";
```

Comment:
Personally, I'd always append a newline after each message. It's way too easy to forget a newline at the end (or `std::endl`) and suddenly find out that the log output looks bad, then adding the newline and recompile everything just for that. That is, for logging code *users*, not having to specify a newline is in most cases a benefit (and in some cases bad, because they explicitly don't want a newline at the end.)

Hmm... actually, I'd rather like to have something like the opposite of `std::endl` - something that tells the stream NOT to append a newline. Just wishful thinking, though.